src/shims/posix/fs.rs

```diff
@@ -1196,6 +1215,85 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         }
     }
 
+    fn linux_readdir64(&mut self, dirp_op: &OpTy<'tcx, Tag>) -> InterpResult<'tcx, Scalar<Tag>> {
+        let this = self.eval_context_mut();
+
+        this.assert_target_os("linux", "readdir64");
+
+        let dirp = this.read_scalar(dirp_op)?.to_machine_usize(this)?;
+
+        // Reject if isolation is enabled.
+        if let IsolatedOp::Reject(reject_with) = this.machine.isolated_op {
+            this.reject_in_isolation("`readdir`", reject_with)?;
+            let eacc = this.eval_libc("EACCES")?;
+            this.set_last_error(eacc)?;
+            return Ok(Scalar::null_ptr(this));
+        }
+
+        let open_dir = this.machine.dir_handler.streams.get_mut(&dirp).ok_or_else(|| {
+            err_unsup_format!("the DIR pointer passed to readdir64 did not come from opendir")
+        })?;
+
+        let entry = match open_dir.read_dir.next() {
+            Some(Ok(dir_entry)) => {
+                let ino64_t_layout = this.libc_ty_layout("ino64_t")?;
+                let off64_t_layout = this.libc_ty_layout("off64_t")?;
+                let c_ushort_layout = this.libc_ty_layout("c_ushort")?;
+                let c_uchar_layout = this.libc_ty_layout("c_uchar")?;
+
+                // If the host is a Unix system, fill in the inode number with its real value.
+                // If not, use 0 as a fallback value.
+                #[cfg(unix)]
+                let ino = std::os::unix::fs::DirEntryExt::ino(&dir_entry);
+                #[cfg(not(unix))]
+                let ino = 0u64;
+
+                let file_type = this.file_type_to_d_type(dir_entry.file_type())?;
+
+                let imms = [
+                    immty_from_uint_checked(ino, ino64_t_layout)?, // d_ino
+                    immty_from_uint_checked(0u128, off64_t_layout)?, // d_off
+                    immty_from_uint_checked(0u128, c_ushort_layout)?, // d_reclen
+                    immty_from_int_checked(file_type, c_uchar_layout)?, // d_type
+                ];
+                let name = dir_entry.file_name();
+                let name_bytes = os_str_to_bytes(&name)?;
+                let name_offset = imms.iter()
+                    .map(|imm| imm.layout.size.bytes())
+                    .sum::<u64>();
+                let size = name_offset
+                    .checked_add(u64::try_from(name_bytes.len()).unwrap())
+                    .unwrap()
+                    .checked_add(1)
```

Comment:
Could you add a comment explaining that this is about the null terminator?

Reply:
Sure!  Or maybe I'll just turn it into a `CString` so I can't forget the NUL terminator